middle_end/augment_closures.ml

```diff
@@ -14,54 +14,90 @@
+  closure_id : Closure_id.t;
+  outside_var : Variable.t;
+}
 
-module Closure_field =
+type closures_in_free_vars = {
+  new_var : Variable.t;
+  start_from : Closure_id.t;
+  outside_var : Variable.t;
+}
+
+type block_in_free_vars = {
+  new_var : Variable.t;
+  outside_var : Variable.t;
+}
+
+module Var_within_closure_field =
   Identifiable.Make (Identifiable.Pair (Variable) (Var_within_closure))
 
+module Closure_field =
+  Identifiable.Make (Identifiable.Pair (Variable) (Closure_id))
+
 module Block_field =
   Identifiable.Make (Identifiable.Pair (Variable) (Numbers.Int))
 
 let freshened_var env v =
   Freshening.apply_variable (E.freshening env) v
 
-let closures_in_variables ~env map acc =
-  Variable.Map.fold (fun inside_var outside_var acc ->
+let closures_in_variables ~env map acc
+      : var_within_closures_in_free_vars Var_within_closure_field.Map.t
+          * closures_in_free_vars Closure_field.Map.t
+          * block_in_free_vars Block_field.Map.t =
+  Variable.Map.fold (fun inside_var outside_var
+        (var_within_closure_acc, closure_acc, block_acc) ->
       let approx = E.find_exn env (freshened_var env outside_var) in
       match A.check_approx_for_closure approx with
       | Ok (value_closure, _approx_var, _approx_symbol,
             value_set_of_closures) ->
-        Var_within_closure.Map.fold (fun bound_var _ (closure_acc, block_acc) ->
-            let new_var =
-              Variable.create (Var_within_closure.unique_name bound_var)
-            in
-            let closure_acc =
-              Closure_field.Map.add (inside_var, bound_var)
-                { new_var; closure_id = value_closure.closure_id; outside_var }
-                closure_acc
-            in
-            closure_acc, block_acc)
-          value_set_of_closures.bound_vars acc
+        let var_within_closure_acc =
+          Var_within_closure.Map.fold
+            (fun bound_var _ acc ->
+              let new_var =
+                Variable.create (Var_within_closure.unique_name bound_var)
+              in
+              Var_within_closure_field.Map.add (inside_var, bound_var)
+                { new_var; closure_id = value_closure.closure_id; outside_var;
+                }
+                acc)
+            value_set_of_closures.bound_vars var_within_closure_acc
+        in
+        let closure_acc =
+          Variable.Map.fold (fun fun_var _ acc ->
+              let new_var = Variable.rename fun_var in
+              let start_from = value_closure.closure_id in
+              let move_to = Closure_id.wrap fun_var in
+              (* For the moment represent all projections of closures as
+                 moves from the original closure ID. *)
+              (* CR-someday mshinwell: consider using the set-of-closures-var
+                 ("_approx_var" above) instead.  See handling in
+                 inline_and_simplify.ml, e.g. simplify_project_var: need
+                 to check the variable is in scope. *)
```

Comment:
Yes normaly a move_closure of project_closure should be simplified.
And if it is not, inline_and_simplify is the place to do that rather than here.